R/pmcmc_nested_parameters.R

```diff
@@ -0,0 +1,488 @@
+      }
+
+      invisible(self)
+    },
+
+    ##' @description Return the names of the parameters
+    names = function() {
+      unlist(private$param_names, use.names = FALSE)
+    },
+
+    ##' @description Return the names of the populations
+    populations = function() {
+      names(private$varied_parameters)
+    },
+
+        ##' @description Return a `data.frame` with information about
+    ##' parameters (name, min, max, discrete, fixed or varied).
+    ##'
+    ##' @param population For parameter sets including varied parameters,
+    ##' `population` specifies which population to summarise. If `NULL` then
+    ##'  returns summary of each population as a list.
+    summary = function(population = NULL) {
+      summarise_population <- function(population) {
+        if (!(population %in% self$populations())) {
+          stop(sprintf("Expected 'population' in %s.",
+            str_collapse(self$populations())))
+        }
+        type <- rep("fixed", length(self$names()))
+        type[self$names() %in% private$param_names$varied] <- "varied"
+        min <- c(r6_private(private$fixed_parameters)$min,
+          r6_private(private$varied_parameters[[population]])$min)
+        max <- c(r6_private(private$fixed_parameters)$max,
+          r6_private(private$varied_parameters[[population]])$max)
+        discrete <- c(r6_private(private$fixed_parameters)$discrete,
+          r6_private(private$varied_parameters[[population]])$discrete)
+
+        data_frame(
+          name = self$names(),
+          min = min,
+          max = max,
+          discrete = discrete,
+          type = type
+        )
+      }
+
+      if (is.null(population)) {
+        sum <- lapply(self$populations(), summarise_population)
+        names(sum) <- self$populations()
+        sum
+      } else {
+        summarise_population(population)
+      }
+    },
+
+    ##' @description Return the initial parameter values as a named numeric
+    ##' vector or named matrix for multiple populations.
+    initial = function() {
+
+      varied <- vapply(private$varied_parameters, function(x) x$initial(),
+        numeric(length(private$param_names$varied)))
+
+      # edge case - 1 varied
+      if (length(private$param_names$varied) == 1) {
+        varied <- t(as.matrix(varied))
+        rownames(varied) <- private$param_names$varied
+      }
+
+      if (is.null(private$fixed_parameters)) {
+        t(varied)
+      } else {
+        fixed <- private$fixed_parameters$initial()
+        t(rbind(
+          matrix(fixed, length(names(fixed)), length(self$populations()),
+            dimnames = list(names(fixed), self$populations())),
+          varied))
+      }
+    },
+
+    ##' @description Compute the prior(s) for a parameter vector/matrix
+    ##'
+    ##' @param theta a parameter matrix with columns in the same order as
+    ##' `$names()` and rows in the same order as `$populations()`.
+    prior = function(theta) {
+      priors <- numeric(nrow(theta))
```

Comment:
This pair of lines uses two sources of truth without checking that they agree (you take the length from nrow(thta) but the names from the number of populations)